File: trustgraph-base/trustgraph/base/parameter_spec.py
from __future__ import annotations

from typing import Any

from . spec import Spec

class Parameter:
    def __init__(self, value: Any) -> None:
        self.value = value
    async def start(self) -> None:
        pass
    async def stop(self) -> None:
        pass
        
class ParameterSpec(Spec):
    def __init__(self, name: str) -> None:
        self.name = name

    def add(self, flow: Any, processor: Any, definition: dict[str, Any]) -> None:

        value = definition.get(self.name, None)

        flow.parameter[self.name] = Parameter(value)

File: ts/packages/base/src/spec/parameter-spec.ts
/**
 * Parameter specification — declares a configuration parameter for a flow.
 *
 * Python reference: trustgraph-base/trustgraph/base/parameter_spec.py
 */

import { Effect } from "effect";
import * as S from "effect/Schema";
import type { SpecRuntimeRequirements } from "./types.js";
import type { Flow, FlowDefinition } from "../processor/flow.js";
import type { PubSubError } from "../errors.js";

declare const ParameterSpecType: unique symbol;

const UnknownParameterSchema: S.Codec<unknown, unknown> = S.Unknown;

export interface ParameterSpec<T = unknown> {
  readonly [ParameterSpecType]?: (_: T) => T;
  readonly name: string;
  readonly schema: S.Codec<T, unknown>;
  readonly addEffect: <Requirements = never>(
    flow: Flow<Requirements>,
    definition: FlowDefinition,
  ) => Effect.Effect<void, PubSubError, SpecRuntimeRequirements | Requirements>;
}

export function makeParameterSpec(name: string): ParameterSpec<unknown>;
export function makeParameterSpec<T>(
  name: string,
  schema: S.Codec<T, unknown>,
): ParameterSpec<T>;
export function makeParameterSpec<T>(
  name: string,
  schema?: S.Codec<T, unknown>,
) {
  const parameterSchema = schema ?? UnknownParameterSchema;
  const addEffect = <Requirements = never>(flow: Flow<Requirements>, definition: FlowDefinition) =>
    Effect.sync(() => {
      const value = definition.parameters?.[name];
      flow.setParameter(name, value);
    });

  return {
    name,
    schema: parameterSchema,
    addEffect,
  };
}
